bip-generic-signmessage.mediawiki

```diff
@@ -0,0 +1,134 @@
+|}
+
+The above is followed by [entries] number of scriptPubKey & signature entries:
+
+{|class="wikitable" style="text-align: center;"
+|-
+!Type
+!Length
+!Name
+!Comment
+|-
+|VarInt||1-8||spklen||Length of scriptPubKey
+|-
+|Uint8*||[spklen]||spk||The scriptPubKey data
+|-
+|VarInt||1-8||scriptsiglen||Number of bytes in scriptSig data
+|-
+|Uint8*||[scriptsiglen]||scriptsig||ScriptSig data
+|-
+|VarInt||1-8||witlen||Number of bytes in witness program data
+|-
+|Uint8*||[witlen]||wit||Witness program
+|}
+
+In some cases, the scriptsig may be empty (scriptsiglen=0).
+
+=== Signing ===
+
+The "Sign" action takes as input a scriptPubKey and a message (e.g. "hello world"). It succeeds or fails.
+
+# FAIL if scriptPubKey already exists in scriptPubKeys set, otherwise insert it<ref><strong>Why track duplicates?</strong> Because a 3-entry proof is not proving 3 scriptPubKeys unless they are all distinct, or unless they are proving different UTXO:s (see Future Extensions)</ref>
+# Derive the private key privkey for the scriptPubKey, or FAIL
+# Define the message pre-image as the sequence "Bitcoin Message:" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD)
+# Let sighash = sha256(sha256(pre-image))
+# Generate a signature sig with privkey=privkey, sighash=sighash
+
+Repeat the above operation for each scriptPubKey, retaining the scriptPubKeys set. As noted, if the same scriptPubKey appears more than once, the sign operation must fail.
+
+=== Verifying ===
+
+The "Verify" action takes as input a standard flags value, a scriptPubKey, a message, a script sig, and a witness program.
+It emits one of INCONCLUSIVE, VALID, INVALID, or ERROR.
+
+# Return ERROR if scriptPubKey already exists in scriptPubKeys set, otherwise insert it
+# If one or more of the standard flags are unknown, return INCONCLUSIVE
+# Define the message pre-image as the sequence "Bitcoin Message:" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD).
+# Let sighash = sha256(sha256(pre-image))
+# Verify Script with flags=standard flags, scriptSig=script sig, scriptPubKey=scriptPubKey, witness=witness program, and sighash=sighash
+# Return VALID if verify succeeds, otherwise return INVALID
+
+Repeat the above operation for each scriptPubKey, retaining the scriptPubKeys set. As noted, if the same scriptPubKey appears more than once, the verify operation must fail with an ERROR.
+
+* If a verification call returns ERROR or INVALID, return ERROR or INVALID immediately, ignoring as yet unverified entries.
+* After all verifications complete, return INCONCLUSIVE if any verification call returned INCONCLUSIVE.
+* Return VALID if and only if every verification returned VALID.
+
+== Future Extensions ==
+
+=== Proof of Funds ===
+
+The specification can be extended to handle proof of funds in the following manner:
```

Comment:
This should just be part of the specification IMO, providing a clear distinction between proof-of-funds (which only verifies an amount, not address(es)) and proof-of-receiving-at-address.

Reply:
I was aiming for a simple and lean proposal and this was only meant as a "you CAN do this, by extending this in the future", but it turned into a more complete proposal than I intended. You're probably right.

Reply:
I've rewritten the proposal to include this as part of the proposal.